2020-12-15-environment-variables-maintenance.md

```diff
@@ -0,0 +1,61 @@
+# Environment Variables Maintenance Proposal
+
+## The Problems
+
+### Documentation
+
+Comment by Ernesto:
+> I believe we need a policy or convention for environment variables. Right now in some projects we are using this policy: If you add a new env variable, you need to document it in both .env.sample and the RELEASE_NOTES file.
+
+### Propagation of the change
+
+It's also hard to maintain: when you pull updates you need to verify the RELEASE_NOTES and compare that with your env variables, you may miss that something changed or that something was added.
+
+### Different requirements for different environments
+
+Not all environment variables are required for development or testing. Currently we have no clear documentation regarding what is optional for development (sometimes a variable inside the .env.sample needs to be removed locally instead of leaving it blank for example).
+
+## Possible solutions
+
+Some suggestions by Ernesto and comments on Twitter were:
+
+### Envied
+
+URL: https://github.com/eval/envied
+
+Pros:
+- To use this gem we need to configure the variables in a `Envfile` (which is just ruby), and then run `ENVied.require` during the app boot.
+- We can add different requirements for different environments on a single file.
+- This can validate many formats.
+- Has a task to check the heroku config against the config file
+- Documentation can be added as comments on the Envfile
+
+Cons:
+- This does NOT support optional env variables (i.e.: variables that can be undefined but if present should have a specific format)
+
+### Prius
+
+URL: https://github.com/gocardless/prius
+
+Pros:
+- We can configure different variables for each environment
+- Supports optional variables
+- Documentation can be added as comments for each variable
+
+Cons:
+- To use this gem we need to call `Prius.load(:some_var)` for all env variables we want to use (we could extract this method calls into a separated file and require that file so we don't pollute the boot files though, not that big of a deal)
+- It can validate only a few types of values
+
+### Custom made gem
+
+We could also create a custom gem or fork any of those if we think it's worth it. The main idea is that, when adding a new env variable, we should be able to add some configuration and documentation related to that variable and check it during boot. That way we can be sure everyone will have the right env variables.
```

Comment:
@arielj I wonder if we could implement a way to use our documentation to be executable code. 

For example: Use `.env.sample` to specify which env variables are required and which are not: 

```
# .env.sample
STRIPE_API_KEY=1234 #required
GOOGLE_ANALYTICS=UA-123
```

Then in a config/initializers/env.rb we could do something like this: 

```
Envied2.check_required!
```

This would read `.env.sample`, find which keys are required, and then check for presence of required keys.

Reply:
I understand we would need to fork dotenv to do that, it doesn't know about comments (# could be a valid char for an environment variable so we would have to come up with a way to define comments)

Reply:
@arielj IIRC, `STRIPE_API_KEY=1234 #required` is valid syntax

Reply:
Yes, that is correct. 

No change to `dotenv` should be necessary: 

```
➜  dash git:(feature/stay-up-to-date) ✗ bundle exec rails console
Running via Spring preloader in process 37761
Loading development environment (Rails 6.1.1)
2.6.6 :001 > ENV['GITHUB_ORGANIZATIONS']
 => "ombulabs"
2.6.6 :002 > exit
➜  dash git:(feature/stay-up-to-date) ✗ cat .env | grep ORG
GITHUB_ORGANIZATIONS="ombulabs" #required
```

It seems that it ignores comments and it loads the right value into the variable.